src/tir/schedule/primitive/blockize_tensorize.cc

```diff
@@ -0,0 +1,638 @@
+  // realize and its surounding loops.
+  Stmt outer_block_body = inner_block_realize;
+  for (const ForNode* loop : inner_loops) {
+    ObjectPtr<ForNode> new_loop = make_object<ForNode>(*loop);
+    new_loop->body = std::move(outer_block_body);
+    outer_block_body = For(new_loop);
+  }
+
+  // Step 4: Generate the outer block and block realize.
+  Block outer_block{/*iter_vars=*/extractor.outer_iter_vars,        //
+                    /*reads=*/new_reads,                            //
+                    /*writes=*/new_writes,                          //
+                    /*name_hint=*/"blockized_" + block->name_hint,  //
+                    /*body=*/std::move(outer_block_body),           //
+                    /*init=*/new_init};
+  BlockRealize outer_block_realize{/*iter_values=*/extractor.outer_bindings,
+                                   /*predicate=*/std::move(predicate),
+                                   /*block=*/std::move(outer_block)};
+  return outer_block_realize;
+}
+
+StmtSRef Blockize(ScheduleState self, const StmtSRef& loop_sref) {
+  const ForNode* loop = TVM_SREF_TO_FOR(loop, loop_sref);
+  arith::Analyzer analyzer;
+
+  // Step 1: Check the loop has a single child BlockRealize on the sref tree.
+  BlockRealize block_realize = CheckGetSingleChildBlockRealizeOnSRefTree(self, loop_sref);
+  Block block = block_realize->block;
+  StmtSRef block_sref = self->stmt2ref.at(block.get());
+
+  // Step 2: Collect loops inside and outside loop_sref.
+  LoopSubspaceCollector collector;
+  collector.Collect(block_sref, loop_sref);
+
+  // Step 3: Calculate subspace division for the inner loops.
+  Array<Array<arith::IterMark>> division =
+      CheckSubspaceDivisible(self->mod, block_realize, collector, &analyzer);
+
+  // Step 4: Generate bindings for the outer block and the inner block based on the result of
+  // the subspace division.
+  BlockizedBindingExtractor extractor;
+  extractor.ExtractBindings(block->iter_vars, division);
+  const PrimExpr& outer_pred = division.back()[0]->extent;
+  const PrimExpr& inner_pred = division.back()[1]->extent;
+
+  // Step 5: Generate the inner block.
+  BlockRealizeNode* inner_block_realize = block_realize.CopyOnWrite();
+  BlockNode* inner_block = inner_block_realize->block.CopyOnWrite();
+  inner_block_realize->iter_values = extractor.inner_bindings;
+  inner_block_realize->predicate = inner_pred;
+  inner_block->iter_vars = extractor.inner_iter_vars;
+  inner_block->init = NullOpt;
+
+  // Step 6: Generate the outer block.
+  BlockRealize outer_realize =
+      GenerateBlockizedOuterBlock(extractor, block, GetRef<BlockRealize>(inner_block_realize),
+                                  collector.inner_loops, outer_pred, &analyzer);
+  // Step 7: Do the actual replacement
+  self->Replace(loop_sref, outer_realize, {{block, GetRef<Block>(inner_block)}});
+
+  // Step 8: Update the cached flags
+  const StmtSRef& outer_block_sref = self->stmt2ref.at(outer_realize->block.get());
```

Comment:
No need to refer to this class, because it's returned by this method

```suggestion
  StmtSRef outer_block_sref = self->stmt2ref.at(outer_realize->block.get());
```